Match Restore Coach Fix

PhlexPlexico opened the pull request:
Per #699 there is an issue during match restores either using `get5_loadbackup` or `!stop` in chat. Once the match is restored, a coach is dropped onto the team, and unable to `.coach` back. Maybe there was a change to the coaching mechanism in CS:GO, but this fix simply allows users to use `.coach` In freeze time, as was intended by the looks of it from the comments?

Thanks!

PhlexPlexico commented:
Moving this to a draft because I would like to make further improvements on how coaches are handled. Mainly the following:

- Use a global ArrayList to keep track of which coach is on which team, that way when a player disconnects and reconnects, they can immediately be thrown back onto the coaching team.
- Add the team coaches to the backup structure, that way during a match restore they will be placed onto the correct coach slot for teams as well.

PhlexPlexico commented:
So player placement on round restore gets to be reaaaaal buggy. However, I think we may be best off just blocking on who can coach on a round restore as well. I'd like to get some more people to test these new features and let me know how it goes, but I don't think I ran into anything too terrible when working with the backup structure.

splewis commented on scripting/get5.sp:
```diff
@@ -107,6 +107,7 @@ char g_FormattedTeamNames[MATCHTEAM_COUNT][MAX_CVAR_LENGTH];
 char g_TeamFlags[MATCHTEAM_COUNT][MAX_CVAR_LENGTH];
 char g_TeamLogos[MATCHTEAM_COUNT][MAX_CVAR_LENGTH];
 char g_TeamMatchTexts[MATCHTEAM_COUNT][MAX_CVAR_LENGTH];
+char g_TeamCoaches[MATCHTEAM_COUNT][64];
```
I believe it's possible for a team to have multiple coaches (i.e. coach + analyst) - I haven't verified that anytime lately though.

PhlexPlexico replied:
Yeah I think this still may be the case, I'll fix this and update it to be an array in coming days!

PhlexPlexico commented:
Okay, to address one of the coach bug issues someone was intermittently having as well, I've introduced `coaches_per_team` in the match config. It's entirely optional, but by default it only allows two coaches per team. 

I've also updated the structs to be plural, since we can allow multiple coaches, and they both load from JSON and KeyValue files. Documentation has been updated to show for it as well.

`g_TeamCoaches` now acts like the player array does, and adds based on if it exists or not. I don't think this should be merged just yet, but everything should be working as intended again. @kubo6472 would you mind testing these changes out if you have the time? I know you were asking about the add coach feature and this should be in there as well.

nickdnk commented on configs/get5/example_match.json:
```diff
@@ -2,6 +2,7 @@
 	"matchid": "example_match",
 	"num_maps": 3,
 	"players_per_team": 1,
```
Might as well make this 5, no?

nickdnk commented:
Fixes https://github.com/splewis/get5/issues/471

nickdnk commented on scripting/get5/teamlogic.sp:
```diff
@@ -354,6 +424,23 @@ public bool AddPlayerToTeam(const char[] auth, MatchTeam team, const char[] name
 }
 
+public bool AddCoachToTeam(const char[] auth, MatchTeam team, const char[] name) {
+  char steam64[AUTH_LENGTH];
```
Move this below the `(team == MatchTeam_TeamSpec)` check.

nickdnk commented on scripting/get5/teamlogic.sp:
```diff
@@ -223,6 +283,24 @@ public MatchTeam GetAuthMatchTeam(const char[] steam64) {
+  }
+
+  if (g_InScrimMode) {
+    return IsAuthOnTeamCoach(steam64, MatchTeam_Team1) ? MatchTeam_Team1 : MatchTeam_Team2;
```
Wouldn't it be safer here to not assume that false == team2 ?

nickdnk commented on scripting/get5/teamlogic.sp:
```diff
@@ -74,21 +85,39 @@ public Action Command_JoinTeam(int client, const char[] command, int argc) {
+public bool CheckIfClientCoaching(int client, MatchTeam team) {
+    // Force user to join the coach if specified by config or reconnect.
+    char clientAuth64[AUTH_LENGTH];
+    GetAuth(client, clientAuth64, AUTH_LENGTH);
```
Move getAuth inside the "if coaching" check. No reason to fetch if it should not be evaluated.

nickdnk replied:
or just return false early at the beginning if g_CoachingEnabledCvar.BoolValue is false
